src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_SourceGenerators.cs

```diff
@@ -26,52 +28,92 @@ internal sealed record SourceGeneratorMap(
     /// of things so that we don't cause source generators to be loaded (and fixed) within VS (which is .net framework
     /// only).
     /// </summary>
-    private static readonly ImmutableArray<(string language, AnalyzerReferencesToSourceGenerators referencesToGenerators, AnalyzerReferencesToSourceGenerators.CreateValueCallback callback)> s_languageToAnalyzerReferencesToSourceGeneratorsMap =
-    [
-        (LanguageNames.CSharp, new(), (static rs => ComputeSourceGenerators(rs, LanguageNames.CSharp))),
-        (LanguageNames.VisualBasic, new(), (static rs => ComputeSourceGenerators(rs, LanguageNames.VisualBasic))),
-    ];
-
-    private static SourceGeneratorMap ComputeSourceGenerators(IReadOnlyList<AnalyzerReference> analyzerReferences, string language)
-    {
-        using var generators = TemporaryArray<ISourceGenerator>.Empty;
-        var generatorToAnalyzerReference = ImmutableDictionary.CreateBuilder<ISourceGenerator, AnalyzerReference>();
+    private static readonly ConditionalWeakTable<ProjectState, SourceGeneratorMap> s_projectStateToSourceGeneratorsMap = new();
 
-        foreach (var reference in analyzerReferences)
-        {
-            foreach (var generator in reference.GetGenerators(language).Distinct())
-            {
-                generators.Add(generator);
-                generatorToAnalyzerReference.Add(generator, reference);
-            }
-        }
-
-        return new(generators.ToImmutableAndClear(), generatorToAnalyzerReference.ToImmutable());
-    }
+    /// <summary>
+    /// Cached information about if a project has source generators or not.  Note: this is distinct from <see
+    /// cref="s_projectStateToSourceGeneratorsMap"/> as we want to be able to compute it by calling over to our OOP
+    /// process (if present) and having it make the determination, without the host necessarily loading generators
+    /// itself.
+    /// </summary>
+    private static readonly ConditionalWeakTable<ProjectState, AsyncLazy<bool>> s_hasSourceGeneratorsMap = new();
```

Comment:
>ConditionalWeakTable [](http://example.com/codeflow?start=28&length=20)

Would it be possible to only have a single CWT and only access the SourceGeneratorMap part of the tuple when OOP?

Reply:
tried, but it started getting very ugly/allocy